### DISCUSSION TITLE
Warn in the doc that onSuccess/onSettled/onError are not called when component unmounts

### DISCUSSION THREAD

**Discussion by ptbrowne** [TanStack/query#1837] [General]
If a component attaches onSuccess via the mutate call on the mutation object returned by useMutation() and the component gets unmounted, for example because the dialog triggering the mutate call is unmounted, onSuccess is not called.
This is understandable but I wish there was a warning in the documentation on this point, I haven't seen anything on this point.

For example here : https://react-query.tanstack.com/guides/updates-from-mutation-responses

Would you be open to a PR ?

```
import { useSnackbar } from 'notistack'

const MyDialog = ({ onClose }) => {
  const mutation = useMutation()
  const { enqueueSnackbar, closeSnackbar } = useSnackbar()
  const onSubmit = ({ values }) => {
    const importing = enqueueSnackbar('Importing')
    onClose()
    mutation.mutate(values, {
      onSuccess: () => {
         closeSnackbar(importing)
         enqueueSnackbar('Import OK')
      }
    }
  }
  return <Dialog />
}
```

Here, the modal is closed before the importing process is finished, on purpose.

To remedy the situation, I used `open={false}` on the MUI dialog, to hide it instead of unmounting, but it was a bit difficult to know why `onSuccess` was not called.

Edit: Actually it's been twice today that I am bitten by this bug. Either I am missing something or could it possibly be a feature request to be able to force lifecycle callbacks to be called even if components that used the mutation was unmounted ?

**Comment by Noitidart** [comment]
Thanks for sharing this, I also got hit by it but with `useQuery`. Did you find anyway for it to be called even if unmounted?

**Reply by ptbrowne** [reply]
I kept my component mounted in the end. Since it was a MUI Dialog, I kept it mounted but hidden through `open=false`.

**Reply by Noitidart** [reply]
Thanks for sharing. What I ended up doing is a catch on the queryFn and moving that global logic to there.

**Comment by TkDodo** [comment]
For mutations, this is mentioned explicitly in the docs because the callbacks on `useMutation` are actually always invoked:

https://tanstack.com/query/v4/docs/guides/mutations#mutation-side-effects
